src/cluster_legacy.c

```diff
@@ -6384,18 +6384,41 @@ sds genClusterInfoString(void) {
         }
     }
 
+    dictIterator *di = dictGetIterator(server.cluster->nodes);
+    dictEntry *de;
+    unsigned nodes_pfail = 0, nodes_fail = 0;
+    while ((de = dictNext(di)) != NULL) {
+        clusterNode *node = dictGetVal(de);
+        if (nodeTimedOut(node)) {
+            nodes_pfail++;
+        }
+        if (nodeFailed(node)) {
+            nodes_fail++;
+        }
+    }
+    dictReleaseIterator(di);
+
+    if (nodes_pfail + nodes_fail > dictSize(server.cluster->nodes) && !server.crashed) {
+        serverLog(LL_WARNING, "Aggregated count of nodes marked as PFAIL and FAIL exceeds the total count of nodes."
+                              "PFAIL nodes: %u, FAIL nodes: %u, total nodes: %lu",
+                  nodes_pfail, nodes_fail, dictSize(server.cluster->nodes));
+        serverAssert(0);
+    }
```

Comment:
Recursive crash? So it means it does assert here sometimes, or only when you changed some logic to make it happen?

I was thinking an assert instead of log, if it's not supposed to happen. If it can happen, then maybe only log the warning. The situation might be harmless so we don't need to panic. Maybe.

```suggestion
    if (!server.crashed) {
        /* Prevent recursive crash */
        serverAssert(nodes_pfail + nodes_fail <= dictSize(server.cluster->nodes));
    }
```

Reply:
I thought of writing this, would it print the actual numbers in that case? Or it would just crash saying assertion failed?

Reply:
Alright, I tested it with a small change. I don't get the variable output captured with the suggested approach.

```c
diff --git a/src/server.c b/src/server.c
index 5affc8d8a..6303d9788 100644
--- a/src/server.c
+++ b/src/server.c
@@ -4698,6 +4698,8 @@ void pingCommand(client *c) {
         return;
     }

+    int x = 10, y = 15;
+    serverAssert((x+y) < 20);
```

```
=== VALKEY BUG REPORT START: Cut & paste starting from here ===
38881:M 03 Apr 2025 21:56:40.416 # === ASSERTION FAILED ===
38881:M 03 Apr 2025 21:56:40.416 # ==> server.c:4702 '(x+y) < 20' is not true
```

Reply:
Yeah. So you need the numbers for debugging. Let's do a warning then and no assert? It's not really necessary to abort I think, since it may be a harmless situation.

I.e. as in your original code, WDYT?

```suggestion
    if (nodes_pfail + nodes_fail > dictSize(server.cluster->nodes)) {
        serverLog(LL_WARNING, "Aggregated count of nodes marked as PFAIL and FAIL exceeds the total count of nodes."
                              "PFAIL nodes: %u, FAIL nodes: %u, total nodes: %lu",
                  nodes_pfail, nodes_fail, dictSize(server.cluster->nodes));
    }
```

Reply:
After you requested the assertion, I find it useful to have. So, that we will take notice of it (if it ever happens). No one looks at the log otherwise. WDYT ?

Reply:
OK, as you wish. :)

Reply:
why we need this? Someone did not handle pfail and fail flags correctly? this like a too-much.

Reply:
@enjoy-binbin Felt good to capture if there is anything wrong on the server side. You think it's very defensive? Shall we clean it up completely?

Reply:
yes, i think it is very defensive, i would vote to remove it.

Reply:
Alright I don't mind removing it. No strong opinion on this.

Reply:
Done.